pkg/graveler/delete_sensor.go

```diff
@@ -0,0 +1,154 @@
+package graveler
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/treeverse/lakefs/pkg/logging"
+)
+
+const (
+	callbackChannelSize = 1000
+	graceDuration       = 30 * time.Second
+)
+
+type DeleteSensorCB func(repositoryID RepositoryID, branchID BranchID, stagingTokenID StagingToken, inGrace bool)
+
+type StagingTokenCounter struct {
+	StagingTokenID StagingToken
+	Counter        int
+}
+
+type stagingTokenData struct {
+	repositoryID   RepositoryID
+	branchID       BranchID
+	stagingTokenID StagingToken
+}
+
+func (s *stagingTokenData) CombinedKey() string {
+	return fmt.Sprintf("%s:%s", s.repositoryID, s.branchID)
+}
+
+type DeleteSensor struct {
+	ctx       context.Context
+	cb        DeleteSensorCB
+	triggerAt int
+	callbacks chan stagingTokenData
+	wg        sync.WaitGroup
+	mutex     *sync.RWMutex
+	// stopped used as flag that the sensor has stopped. stop processing CountDelete.
+	stopped                int32
+	graceDuration          time.Duration
+	branchTombstoneCounter map[string]*StagingTokenCounter
+}
+
+type DeleteSensorOpts func(s *DeleteSensor)
+
+func WithCBBufferSize(bufferSize int) DeleteSensorOpts {
+	return func(s *DeleteSensor) {
+		s.callbacks = make(chan stagingTokenData, bufferSize)
+	}
+}
+
+func WithGraceDuration(d time.Duration) DeleteSensorOpts {
+	return func(s *DeleteSensor) {
+		s.graceDuration = d
+	}
+}
+func NewDeleteSensor(ctx context.Context, triggerAt int, cb DeleteSensorCB, opts ...DeleteSensorOpts) *DeleteSensor {
+	ds := &DeleteSensor{
+		ctx:                    ctx,
+		cb:                     cb,
+		triggerAt:              triggerAt,
+		stopped:                0,
+		graceDuration:          graceDuration,
+		mutex:                  &sync.RWMutex{},
+		branchTombstoneCounter: make(map[string]*StagingTokenCounter),
+		callbacks:              make(chan stagingTokenData, callbackChannelSize),
+	}
+	for _, opt := range opts {
+		opt(ds)
+	}
+	ds.wg.Add(1)
+	go ds.processCallbacks()
+	return ds
+}
+
+func (s *DeleteSensor) isStopped() bool {
+	return atomic.LoadInt32(&s.stopped) == 1
+}
+
+func (s *DeleteSensor) triggerTombstone(ctx context.Context, st stagingTokenData) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	combinedKey := st.CombinedKey()
+	stCounter, ok := s.branchTombstoneCounter[combinedKey]
+	if !ok {
+		stCounter = &StagingTokenCounter{
+			StagingTokenID: st.stagingTokenID,
+			Counter:        1,
+		}
+		s.branchTombstoneCounter[combinedKey] = stCounter
+		return
+	}
+	if stCounter.StagingTokenID != st.stagingTokenID {
+		stCounter.StagingTokenID = st.stagingTokenID
+		stCounter.Counter = 1
+		return
+	}
```

Comment:
Why drop the count of the sealed token? Its entries will _still_ need to be skipped.

I think that the assumption here is that if the token was sealed then a commit is in progress, meaning we will soon be rid of those entries.  But then please document this decision -- we may have other reasons to seal in future.

Reply:
This was done to:
1. Avoid the need for exposing a reset method and calling across the code (for now)
2. An "easy" way to reset cross lakeFS instances when a `commit` or `compaction` is done.

It isn't the best way and has many downsides, but I think it will do for the log implementation, and will give us valid information for how we would like to go on

Adding documentation

Reply:
You might have a weird order of staging token for a branch, a transition of staging tokens might look like:
`st1,st1,st1,st2,st1,st2,st2,st2`. How will this effect the sensor logic?

Reply:
I'm fine with opening an issue to refactor this and revisit at a later date.

At such a later date, I would probably try to fill out this sketch: Keep a sync.Map of branch or staging token triggers (I'm fine with either, because we don't care about losing a few counts -- in any case if we need to compact more than once a second, then we need to do something better than compaction!).  Whenever we have to, atomically increment the counter on the trigger.  If the count-after-increment is (exactly) 1000:
- If we count per branch, atomically decrement it by 1000 and trigger compaction or whatever.
- If we count per staging token, atomically replace the staging token trigger, and we're happy to lose the excess counts!